datatypes/src/raster/raster_tile.rs

```diff
@@ -1,96 +1,188 @@
 use num_traits::AsPrimitive;
 use serde::{Deserialize, Serialize};
 
-pub type RasterTile2D<T> = RasterTile<Dim2D, T>;
-pub type RasterTile3D<T> = RasterTile<Dim3D, T>;
+pub type RasterTile2D<T> = RasterTile<ArrayShape2D, T>;
+pub type RasterTile3D<T> = RasterTile<ArrayShape3D, T>;
 
 /// A `RasterTile2D` is the main type used to iterate over tiles of 2D raster data
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 pub struct RasterTile<D, T>
 where
+    D: GridSize + GridSpaceToLinearSpace,
     T: Pixel,
 {
     pub time: TimeInterval,
-    pub tile: TileInformation,
-    pub data: BaseRaster<D, T, Vec<T>>,
+    pub tile_position: GridIdx2D,
+    pub global_geo_transform: GeoTransform,
+    pub grid_array: GridArray<D, T>,
 }
 
 impl<D, T> RasterTile<D, T>
 where
     T: Pixel,
+    D: GridSize + GridSpaceToLinearSpace + Clone,
 {
-    /// create a new `RasterTile2D`
-    pub fn new(time: TimeInterval, tile: TileInformation, data: BaseRaster<D, T, Vec<T>>) -> Self {
-        Self { time, tile, data }
+    /// create a new `RasterTile`
+    pub fn new_with_tile_info(
+        time: TimeInterval,
+        tile_info: TileInformation,
+        data: GridArray<D, T>,
+    ) -> Self {
+        // TODO: assert, tile information xy size equals the data xy size
+        Self {
+            time,
+            tile_position: tile_info.global_tile_position,
+            global_geo_transform: tile_info.global_geo_transform,
+            grid_array: data,
+        }
+    }
+
+    /// create a new `RasterTile`
+    pub fn new(
+        time: TimeInterval,
+        tile_position: GridIdx2D,
+        global_geo_transform: GeoTransform,
+        data: GridArray<D, T>,
+    ) -> Self {
+        Self {
+            time,
+            tile_position,
+            global_geo_transform,
+            grid_array: data,
+        }
+    }
+
+    /// create a new `RasterTile`
+    pub fn new_without_offset(
+        time: TimeInterval,
+        global_geo_transform: GeoTransform,
+        data: GridArray<D, T>,
+    ) -> Self {
+        Self {
+            time,
+            tile_position: [0, 0].into(),
+            global_geo_transform,
+            grid_array: data,
+        }
     }
 
     /// Converts the data type of the raster tile by converting its inner raster
     pub fn convert<To>(self) -> RasterTile<D, To>
     where
-        D: GridDimension,
+        D: GridSize + GridSpaceToLinearSpace,
         To: Pixel + FromPrimitive<T>,
         T: AsPrimitive<To>,
     {
-        RasterTile::new(self.time, self.tile, self.data.convert())
+        RasterTile::new(
+            self.time,
+            self.tile_position,
+            self.global_geo_transform,
+            self.grid_array.convert_dtype(),
+        )
+    }
+
+    pub fn grid_dimension(&self) -> D {
+        self.grid_array.shape.clone()
+    }
+
+    pub fn tile_offset(&self) -> GridIdx2D {
+        self.tile_position
+    }
+
+    pub fn tile_information(&self) -> TileInformation {
```

Comment:
why is the TileInformation built on demand on not a field of the RasterTile?

Reply:
this would store the size of the grid twice.